lib/spack/llnl/util/tty/log.py

```diff
@@ -426,45 +440,71 @@ def _writer_daemon(self, stdin):
         istreams = [in_pipe, stdin] if stdin else [in_pipe]
 
         log_file = self.log_file
+
+        def handle_write(force_echo):
+            # Handle output from the with block process.
+            # If we arrive here it means that in_pipe was
+            # ready for reading : it should never happen that
+            # line is false-ish
+            line = in_pipe.readline()
+            if not line:
+                return (True, force_echo)  # break while loop
+
+            # find control characters and strip them.
+            controls = control.findall(line)
+            line = re.sub(control, '', line)
+
+            # Echo to stdout if requested or forced
+            if echo or force_echo:
+                try:
+                    import termios
```

Comment:
No, the import statement needs to be part of the `try/except` block.

Reply:
A couple things:

* it can move - but you'd need to store the result of whether that succeeded in a module-level variable
* once that is done, IMO it's clearer to skip `with keyboard_input(stdin):` entirely when `termios_unavailable` since AFAIK it serves no purpose when that module is not available

Reply:
The `llnl.util` modules are intended to be useful outside of Spack. Considering that, I think it is preferable to do the import checking for `keyboard_input` inside of the context manager, rather than outside (as I would then need to conditionally define the class, and users would then need to put try/except blocks around their own import code. IMO it is better to have a class that does keyboard input when possible, and fails transparently.

The particular code you commented on actually is not inside of the keyboard_input class.

In general, I think it is clearer to put a try/except around a block of code containing an import than to pass boolean flags around.

Reply:
For me the question is: what should an object do when it can’t perform its intended operation? In every case where I have encountered that, I have either thrown an exception or added functionality to report that to the user (and allow them to check). I haven't seen examples of objects which silently decide not to do what the user requests.

I think your objections are valid for my suggested approach although I would still consider my suggested change an improvement - in particular since nothing outside of this module uses `keyboard_input`. Could you think of another way to get that behavior? Possibly `keyboard_input.__enter__` could itself throw an exception if `termios` cannot be imported (in that case something more specific than `Exception`).

Reply:
Ok, my new plan is to change `keyboard_input` to `_keyboard_input` and only conditionally define the class when `termios` is available.

Reply:
See my review comments.  I agree with this:

> In general, I think it is clearer to put a try/except around a block of code containing an import than to pass boolean flags around.

I think this is a misinterpretation of what `keyboard_input` is supposed to do:

> I haven't seen examples of objects which silently decide not to do what the user requests.

In particular, it marks a region where we can get keyboard input from the user.  It's best effort -- we do not *NEED* `termios`, and it is best effort for the particular platform. If we do not have `termios`, the user can type `v` and then `RETURN` to get verbose ouput -- it's really just about disabling line buffering where possible (which is most places).

I agree that if there were a *failure* case here that we might want to make that reportable, but this thing is really best effort.  It's supposed to be a transparent context manager.

it is *very* unusual to catch an exception from the entry point of a context manager.  If you want something to act on what happens inside the context manager, you can yield it, e.g.:

```python
with keyboard_input(...) as line_buffered:
    if line_buffered:
        # do something different if needed
```

And then the caller can ignore the whole `as` clause if they want to (as we do).  But I don't really consider it a failure if `termios` isn't available.